docs/design/2021-03-30-small-table-caching.md

```diff
@@ -0,0 +1,174 @@
+
+- `tid` records the ID of the table being cached
+- `lock_type` decides whether the current state is read locked or write locked etc
+- `lease` is used for the read locks lease renewal, and also for cleanup the write locks
+- `oldReadLease` is necessary when the `lock_type` is 'INTEND', it stores the old read lease
+
+LockForRead() corresponds to this operation:
+
+```
+update table_cache_meta set lock ='READ', lease = %? where  tid = %? and lock !='WRITE'
+```
+
+RenewLease() corresponds to this operation:
+
+```
+update table_cache_meta set lease = %? where tid = %? and lock ='READ'
+```
+
+LockForWrite() corresponds to this operation:
+
+```
+update table_cache_meta set lock ='INTEND' and oldLease = %? and lease = %?' where tid = %?
+sleep() // wait read lease gone
+update table_cache_meta set lock ='WRITE' and lease = %?' where tid = %?
+```
+
+### Caching and writing process
+
+The way to store the in-memory cached of the table data is the same with the [temporary table](2021-04-20-temporary-table.md).
+In short, the KV data of the table is loaded from TiKV, and is stored in a MemBuffer. To read that data, an extra UnionScan executor is used.
+
+
+A `CachedTable` struct will be introduced. It inherits and overloads the `Table` interface. `loadSchema` will periodically load the table information on schema change. For a cached table, the `CachedTable` will be reconstructed.
+
+`CachedTable` implements the `Table` interface, and overloads the `AddRecord` method. When `AddRecord` is called, the lock meta information needs to be checked first. This operation needs to acquire the "write" lock, that is to say, make the `lock_type` field of the meta table to 'WRITE'. After confirming that the lease time of the previous 'READ' lock has expired, the operation can be executed.
+
+The read operation flow diagram:
+
+![](./imgs/cached-table-read.png)
+
+
+The write operation flow diagram:
+
+![](./imgs/cached-table-write.png)
+
+
+### How to enable this feature
+
+The syntax of OceanBase's replicated table is to add the `DUPLICATE_SCOPE` option after the CREATE TABLE statement.
+
+I propose to use "ALTER TABLE t [CACHE|NOCACHE]" for the syntax. This is a switch that can be turned on or off. A user can decide to use it by themself. The write performance will become very bad after caching.
+
+"ALTER TABLE t CACHE" is a DDL operation. Assuming that all the TiDB instances know that a table is not cached, there is no correctness problem; Assuming that all tidb instances make a consensus that a table is cached, the correctness is guaranteed by the locking algorithm mentioned before. However, during the DDL operation, if some TiDB instances think that the cache is enabled, while some other instances think that the cache is not enabled, there would be a correctness issue. 
+
+To address that problem, an intermediate `Switching` state is introduced, the schema change process is similar to the LOCK TABLE statement's implementation:
```

Comment:
I assume that because there is a `Switching State` there is no risk that using an older server binary in the cluster will lead to correctness issues, because Enabled can never be reached?

However, there is a second issue:
- Once the enabled state is entered, how do you prevent that an older version of TiDB is not started, and causes correctness issues by not respecting the flag?

Reply:
I never assume different version tidb binaries run together.
The only scene that there are both new and old version tidb server binary coexist I can come up with is the rolling upgrade/downgrade scenario.

Rolling upgrade should be almost safe, because this new feature is not used yet.
While rolling downgrade is not supported, there is no correctness guarantee in that case.

The DDL schema change protocol is used to ensure the safety of the DDL operation, and here it can't (and is not designed to) handle the rolling upgrade operation.

Reply:
> how do you prevent that an older version of TiDB is not started, and causes correctness issues

I can't prevent that.